i2c/zynq-7000/libzynq7000-i2c.c

```diff
@@ -0,0 +1,509 @@
+/*
+ * Phoenix-RTOS
+ *
+ * Zynq-7000 I2C driver
+ *
+ * Copyright 2022 Phoenix Systems
+ * Author: Hubert Buczynski
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include <sys/mman.h>
+#include <sys/msg.h>
+#include <sys/platform.h>
+#include <sys/threads.h>
+#include <sys/interrupt.h>
+#include <sys/mman.h>
+
+#include <stdint.h>
+#include <unistd.h>
+#include <errno.h>
+#include <i2c.h>
+
+#include <phoenix/arch/zynq7000.h>
+#include <board_config.h>
+
+/* I2C registers */
+#define I2C_CR        0
+#define I2C_SR        1
+#define I2C_ADDR      2
+#define I2C_DATA      3
+#define I2C_ISR       4
+#define I2C_TRANS_SZ  5
+#define I2C_SLV_PAUSE 6
+#define I2C_TIMEOUT   7
+#define I2C_IMR       8
+#define I2C_IER       9
+#define I2C_IDR       10
+
+#define I2C_FIFO_DEPTH     16
+#define I2C_TRANS_SIZE_MAX 255
+
+/* MIO I2C pins configuration */
+#define MIO_SDA 0x1240
+#define MIO_SCL 0x1240
+
+
+typedef struct {
+	unsigned int id;  /* I2C controller ID */
+	unsigned int irq; /* I2C controller IRQ */
+	addr_t paddr;     /* I2C controller base physical address */
+
+	struct {              /* I2C MIO pins configuration */
+		int pin;          /* MIO pin */
+		unsigned int cfg; /* MIO pin configuration */
+	} pins[2];            /* I2C pins, CLK, SDA */
+} i2c_info_t;
+
+
+/* device configuration based on board_config */
+static const i2c_info_t devsInfo[] = {
+	{
+		.id = 0,
+		.irq = 57,
+		.paddr = 0xe0004000,
+		.pins = {
+			{ I2C0_SDA, MIO_SDA },
+			{ I2C0_SCL, MIO_SCL },
+		}
+	},
+	{
+		.id = 1,
+		.irq = 80,
+		.paddr = 0xe0005000,
+		.pins = {
+			{ I2C1_SDA, MIO_SDA },
+			{ I2C1_SCL, MIO_SCL },
+		}
+	}
+};
+
+
+static struct {
+	unsigned int devID;
+	volatile uint32_t *base; /* I2C registers base address */
+	int initialized;
+
+	handle_t lock;           /* I2C IRQ mutex */
+	handle_t cond;           /* I2C IRQ cond */
+	handle_t inth;           /* I2C IRQ handle */
+
+	volatile uint32_t st;
+} i2c = { 0 };
+
+
+static int i2c_isr(unsigned int n, void *arg)
+{
+	i2c.st = *(i2c.base + I2C_ISR);
+
+	/* clear status register */
+	*(i2c.base + I2C_ISR) = i2c.st;
+
+	/* transfer complete */
+	if (i2c.st & 0x1) {
+		*(i2c.base + I2C_IDR) |= 0x1;
+	}
+	/* data irq */
+	else if (i2c.st & (1 << 1)) {
+		*(i2c.base + I2C_IDR) |= (1 << 1);
+	}
+	/* transfer NACK */
+	else if (i2c.st & (1 << 2)) {
+		*(i2c.base + I2C_IDR) |= (1 << 2);
+	}
+	/* transfer timeout */
+	else if (i2c.st & (1 << 3)) {
+		*(i2c.base + I2C_IDR) |= (1 << 3);
+	}
+	/* transmit overflow */
+	else if (i2c.st & (1 << 6)) {
+		*(i2c.base + I2C_IDR) |= (1 << 7);
+	}
```

Comment:
It is ok?
Shouldn't it be:
```suggestion
	else if (i2c.st & (1 << 6)) {
		*(i2c.base + I2C_IDR) |= (1 << 6);
	}
```

Reply:
Thanks for pointing this out. copy-paste mistake.